File: fortran/parameter_container.f90
module parameter_container_class

  use abstract_parameter_class, only : abstract_parameter
  use basis_helper            , only : basis_degrees
  use quadrature_helper       , only : tensor_product
  
  implicit none

  type param_ptr
     class(abstract_parameter), pointer :: p => null()
  end type param_ptr

  type :: parameter_container

     integer :: num_parameters = 0

     type(param_ptr), dimension(5) :: plist
     integer        , dimension(5) :: param_maxdeg = 0
     ! integer        , dimension(5) :: param_nqpts = 0

     integer, allocatable :: dindex(:,:)
     real(8), allocatable :: zz(:,:), yy(:,:), ww(:)
     
     integer :: num_basis_terms = 0
     logical :: basis_initialized = .false.

     integer :: num_quadrature_points = 0
     logical :: quadrature_initialized = .false.
     
   contains

     procedure :: initialize_basis
     generic   :: basis => basis_term, basis_given_degrees
     
     procedure :: initialize_quadrature
     procedure :: get_quadrature
     
     procedure :: add
     procedure :: get_num_basis_terms
     procedure :: get_num_quadrature_points
     
     procedure, private :: basis_term, basis_given_degrees
     
  end type parameter_container

contains
  
  subroutine initialize_quadrature(this, pnqpts)

    class(parameter_container), intent(inout) :: this
    integer                   , intent(in) :: pnqpts(:)

    real(8) :: zp(this % num_parameters, maxval(pnqpts))
    real(8) :: yp(this % num_parameters, maxval(pnqpts))
    real(8) :: wp(this % num_parameters, maxval(pnqpts))
    integer :: p
    
    if (allocated(this % zz)) deallocate(this % zz)
    if (allocated(this % yy)) deallocate(this % yy)
    if (allocated(this % ww)) deallocate(this % ww) 
        
    do p = 1, this % num_parameters
       call this % plist (p) % p % quadrature(&
            & pnqpts(p), &
            & zp(p, 1 : pnqpts(p)), &
            & yp(p, 1 : pnqpts(p)), &
            & wp(p, 1 : pnqpts(p)) &
            & )
    end do

    call tensor_product(&
         & pnqpts, &
         & zp, yp, wp, &
         & this % zz, this % yy, this % ww)
    
    if (allocated(this % zz)) then 
       this % num_quadrature_points = size(this % zz, dim = 2)
    else
       print *, "quadrature points not allocated"
    end if
    
  end subroutine initialize_quadrature

  subroutine initialize_basis(this, pmax)

    class(parameter_container), intent(inout) :: this
    integer, intent(in) :: pmax(:)
    integer :: nterms, k

    !  Set the max degree for parameters
    this % param_maxdeg (1 : this % num_parameters) = pmax(:)

    ! generate and store a set of indices
    this % dindex = basis_degrees(this % param_maxdeg(1 : this % num_parameters))

    ! setup number of terms (assume tensor basis)
    this % num_basis_terms = size(this % dindex, dim=2)

    ! initialization flag true
    this % basis_initialized = .true.
    
  end subroutine initialize_basis

  impure subroutine add(this, param)

    class(parameter_container), intent(inout)      :: this
    class(abstract_parameter) , intent(in), target :: param

    if (this % num_parameters .gt. 5)  then
       print *, 'container full -- cannot add more parameters'
    end if

    this % num_parameters = this % num_parameters + 1
    this % plist (param % get_parameter_id()) % p => param        
    call this % plist(this % num_parameters) % p % print()

  end subroutine add

  pure integer function get_num_quadrature_points(this) result(nqpts)

    class(parameter_container) , intent(in) :: this

    nqpts = this % num_quadrature_points

  end function get_num_quadrature_points

  pure integer function get_num_basis_terms(this) result(nterms)
    
    class(parameter_container) , intent(in) :: this
    
    nterms = this % num_basis_terms
    
  end function get_num_basis_terms
  
  real(8) function basis_term(this, k, z) result(psi)

    class(parameter_container) , intent(in) :: this
    real(8)                    , intent(in) :: z(:)
    integer                    , intent(in) :: k
    integer :: ii, nvars
    
    associate(d=>this % dindex(:,k))
      psi = 1.0d0    
      do ii = 1, this % num_parameters
         psi = psi * this % plist(ii) % p % basis(z(ii), d(ii))
      end do
    end associate

  end function basis_term

  real(8) function basis_given_degrees(this, d, z) result(psi)

    class(parameter_container) , intent(in) :: this
    real(8)                    , intent(in) :: z(:)
    integer                    , intent(in) :: d(:)   
    integer :: ii

    psi = 1.0d0    
    do ii = 1, this % num_parameters
       psi = psi * this % plist(ii) % p % basis(z(ii), d(ii))       
    end do

  end function basis_given_degrees
  
  subroutine get_quadrature(this, q, z, y, w)

    class(parameter_container) , intent(in)    :: this
    integer                    , intent(in)    :: q
    real(8)                    , intent(inout) :: z(:), y(:)
    real(8)                    , intent(inout) :: w

    z = this % zz(:,q)
    y = this % yy(:,q)
    w = this % ww(q)

  end subroutine get_quadrature

end module parameter_container_class

File: cpp/ParameterContainer.cpp
#include"ParameterContainer.h"

using namespace std;

ParameterContainer::ParameterContainer(){
  this->tnum_parameters = 0;
}
ParameterContainer::~ParameterContainer(){
  // Clear information about parameters
  if (param_max_degree){ 
    delete [] param_max_degree; 
  };

  // Degree of kth basis entry
  for (int k = 0; k < this->getNumBasisTerms(); k++){
    delete [] this->dindex[k];
  };
  delete [] this->dindex;

  // Deallocate quadrature information
  for (int i = 0; i < this->getNumParameters(); i++){
    delete [] Z[i];
    delete [] Y[i];
  }
  delete [] Z;
  delete [] Y;
  delete [] W;
}

/*
  Add the parameter into  collection
*/
void ParameterContainer::addParameter(AbstractParameter *param){
  if (this->tnum_parameters > 4){
    printf("Warning: Parameter container is full\n");
  }
  this->pmap.insert(std::pair<int, AbstractParameter*>(param->getParameterID(), param));
  this->tnum_parameters++;
}

int ParameterContainer::getNumBasisTerms(){
  return this->tnum_basis_terms;
}

int ParameterContainer::getNumParameters(){
  return this->tnum_parameters;
}

int ParameterContainer::getNumQuadraturePoints(){
  return this->tnum_quadrature_points;
}

void  ParameterContainer::initializeBasis(const int *pmax){
  int nvars = this->getNumParameters();
  
  // Copy over the max degrees
  param_max_degree = new int[nvars];
  for (int k = 0; k < nvars; k++){
    param_max_degree[k] = pmax[k];
  }

  // Number of terms from tensor product
  int nterms = 1;
  for (int i = 0; i < nvars; i++){
    nterms *= 1 + pmax[i];
  }

  // Allocate space for storing degree set
  this->dindex = new int*[nterms];
  for (int k = 0; k < nterms; k++){
    this->dindex[k] = new int[nvars];
  }
  
  // Generate and store a set of indices
  this->bhelper->basisDegrees(nvars, pmax, 
                              &this->tnum_basis_terms, 
                              this->dindex);

  if (this->tnum_basis_terms != nterms){
    printf("Error initializing basis\n");
  };
}

void ParameterContainer::initializeQuadrature(const int *nqpts){  
  const int nvars = getNumParameters();
  int totquadpts = 1;
  for (int i = 0; i < nvars; i++){
    totquadpts *= nqpts[i];
  }
  this->tnum_quadrature_points = totquadpts;
  
  // Get the univariate quadrature points from parameters
  scalar **y = new scalar*[nvars];
  scalar **z = new scalar*[nvars];
  scalar **w = new scalar*[nvars];
  for (int i = 0; i < nvars; i++){
    z[i] = new scalar[nqpts[i]];
    y[i] = new scalar[nqpts[i]];
    w[i] = new scalar[nqpts[i]];    
  }
  map<int,AbstractParameter*>::iterator it;
  for (it = this->pmap.begin(); it != this->pmap.end(); it++){
    int pid = it->first;
    it->second->quadrature(nqpts[pid], z[pid], y[pid], w[pid]);
  }

  // Compute multivariate quadrature and store
  Z = new scalar*[nvars];
  Y = new scalar*[nvars];
  W = new scalar[totquadpts];  
  for (int i = 0; i < nvars; i++){
    Z[i] = new scalar[totquadpts];
    Y[i] = new scalar[totquadpts];    
  }

  // Find tensor product of 1d rules
  qhelper->tensorProduct(nvars, nqpts, z, y, w, Z, Y, W);
  
  // Deallocate space
  for (int i = 0; i < nvars; i++){
    delete [] z[i];
    delete [] y[i];
    delete [] w[i];
  }
  delete [] y;
  delete [] z;
  delete [] w;
}

/*
 */
scalar ParameterContainer::quadrature(int q, scalar *zq, scalar *yq){
  const int nvars = getNumParameters(); 
  for (int i = 0; i < nvars; i++){
    zq[i] = this->Z[i][q];
    yq[i] = this->Y[i][q];
  }
  return this->W[q];
}

/*
  Evaluate the k-the basis function at point "z"
*/
scalar ParameterContainer::basis(int k, scalar *z){
  scalar psi = 1.0;
  map<int,AbstractParameter*>::iterator it;
  for (it = this->pmap.begin(); it != this->pmap.end(); it++){
    int pid = it->first;
    psi *= it->second->basis(z[pid], this->dindex[k][pid]);
  }
  return psi;
}

void ParameterContainer::getBasisParamDeg(int k, int *degs) {
  const int nvars = getNumParameters(); 
  for (int i = 0; i < nvars; i++){
    degs[i] = this->dindex[k][i];
  }
}

void ParameterContainer::getBasisParamMaxDeg(int *pmax) {
  map<int,AbstractParameter*>::iterator it;
  for (it = this->pmap.begin(); it != this->pmap.end(); it++){
    int pid = it->first;
    int dmax = it->second->getMaxDegree();
    pmax[pid] = dmax;
  }
}

/*
  Default initialization 
*/
void ParameterContainer::initialize(){
  const int nvars = getNumParameters();
  int nqpts[nvars];
  int pmax[nvars];    
  map<int,AbstractParameter*>::iterator it;
  for (it = this->pmap.begin(); it != this->pmap.end(); it++){
    int pid = it->first;
    int dmax = it->second->getMaxDegree();
    pmax[pid] = dmax;
    nqpts[pid] = dmax + 1;
  }
  this->initializeBasis(pmax);  
  this->initializeQuadrature(nqpts);
}
